backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/impl/PurchaseOrderLineModularContractHandler.java

```diff
@@ -0,0 +1,156 @@
+import de.metas.contracts.modular.log.LogEntryReverseRequest;
+import de.metas.contracts.modular.log.ModularContractLogDAO;
+import de.metas.contracts.modular.settings.ModularContractSettings;
+import de.metas.contracts.modular.settings.ModularContractSettingsDAO;
+import de.metas.contracts.modular.settings.ModularContractType;
+import de.metas.contracts.modular.settings.ModularContractTypeId;
+import de.metas.contracts.modular.settings.ModuleConfig;
+import de.metas.lang.SOTrx;
+import de.metas.money.CurrencyId;
+import de.metas.money.Money;
+import de.metas.order.IOrderDAO;
+import de.metas.order.OrderId;
+import de.metas.order.OrderLineId;
+import de.metas.organization.IOrgDAO;
+import de.metas.organization.InstantAndOrgId;
+import de.metas.organization.OrgId;
+import de.metas.product.ProductId;
+import de.metas.quantity.Quantity;
+import de.metas.uom.IUOMDAO;
+import de.metas.uom.UomId;
+import de.metas.util.Services;
+import lombok.NonNull;
+import org.adempiere.util.lang.impl.TableRecordReference;
+import org.adempiere.warehouse.WarehouseId;
+import org.compiere.model.I_C_Order;
+import org.compiere.model.I_C_OrderLine;
+import org.compiere.model.I_C_UOM;
+import org.springframework.stereotype.Component;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+@Component
+public class PurchaseOrderLineModularContractHandler implements IModularContractTypeHandler<I_C_OrderLine>
+{
+	private final IOrderDAO orderDAO = Services.get(IOrderDAO.class);
+	private final IOrgDAO orgDAO = Services.get(IOrgDAO.class);
+	private final IUOMDAO uomDAO = Services.get(IUOMDAO.class);
+	private final IFlatrateDAO flatrateDAO = Services.get(IFlatrateDAO.class);
+	public final ModularContractLogDAO modularContractLogDAO;
+	public final ModularContractSettingsDAO modularContractSettingsDAO;
+
+	public PurchaseOrderLineModularContractHandler(@NonNull final ModularContractLogDAO modularContractLogDAO,
+			@NonNull final ModularContractSettingsDAO modularContractSettingsDAO)
+	{
+		this.modularContractLogDAO = modularContractLogDAO;
+		this.modularContractSettingsDAO = modularContractSettingsDAO;
+	}
+
+	@Override
+	public boolean probablyAppliesTo(@NonNull final Object model)
+	{
+		return model.getClass().isAssignableFrom(I_C_OrderLine.class);
+	}
+
+	@Override
+	@NonNull
+	public Optional<LogEntryCreateRequest> createLogEntryCreateRequest(@NonNull final I_C_OrderLine model, @NonNull final FlatrateTermId flatrateTermId)
+	{
+		final ModularContractSettings modularContractSettings = modularContractSettingsDAO.getByFlatrateTermIdOrNull(flatrateTermId);
+		if (modularContractSettings == null)
+		{
+			return Optional.empty();
+		}
+
+		final I_C_Order order = orderDAO.getById(OrderId.ofRepoId(model.getC_Order_ID()));
+		final Optional<ModularContractTypeId> modularContractTypeId = modularContractSettings.getModuleConfigs()
+				.stream()
+				.map(ModuleConfig::getModularContractType)
```

Comment:
u might want to refactor this using the ModuleConfig.isMatchingClassName method I've suggested to introduce above.

Reply:
yup.